flow/layers/physical_model_layer.h

```diff
@@ -35,12 +39,83 @@ class PhysicalModelLayer : public ContainerLayer {
 #endif  // defined(OS_FUCHSIA)
 
  private:
-  SkRRect rrect_;
+  std::unique_ptr<PhysicalLayerShape> shape_;
   float elevation_;
   SkColor color_;
   SkScalar device_pixel_ratio_;
 };
 
+// Common interface for the shape operations needed by PhysicalModelLayer.
+class PhysicalLayerShape {
+ public:
+  virtual const SkRect& getBounds() const = 0;
+  virtual const SkPath getPath() const = 0;
+  virtual void clipCanvas(SkCanvas& canvas) const = 0;
+  virtual bool isRect() const = 0;
+#if defined(OS_FUCHSIA)
+  virtual const SkRRect& getFrameRRect() const = 0;
+#endif  // defined(OS_FUCHSIA)
+};
+
+class PhysicalLayerRRect : public PhysicalLayerShape {
+ public:
+  PhysicalLayerRRect(const SkRRect& rrect) { rrect_ = rrect; }
+
+  // |flow::PhysicalLayerShape|
+  const SkRect& getBounds() const { return rrect_.getBounds(); }
+
+  // |flow::PhysicalLayerShape|
+  const SkPath getPath() const;
+
+  // |flow::PhysicalLayerShape|
+  void clipCanvas(SkCanvas& canvas) const { canvas.clipRRect(rrect_, true); }
+
+  // |flow::PhysicalLayerShape|
+  bool isRect() const { return rrect_.isRect(); }
+
+#if defined(OS_FUCHSIA)
+  // |flow::PhysicalLayerShape|
+  const SkRRect& getFrameRRect() const { return rrect_; }
+#endif  // defined(OS_FUCHSIA)
+
+ private:
+  SkRRect rrect_;
+};
+
+class PhysicalLayerPath : public PhysicalLayerShape {
```

Comment:
Nit: Can we make these subclasses final?

Reply:
done